### PR TITLE
[DROOLS-5861] Fixing KieBase/Classloader usage

### PR DESCRIPTION
**Pull request by gitgabrio**
@danielezonca @mariofusco @jiripetrlik

See https://issues.redhat.com/browse/DROOLS-5861

Depends on
https://github.com/kiegroup/droolsjbpm-knowledge/pull/492

Needed by:
https://github.com/kiegroup/kogito-runtimes/pull/973

### CODE REVIEW

File: drools-core/src/main/java/org/drools/core/command/runtime/pmml/ApplyPmmlModelCommand.java [hunk 1]
```diff
@@ -167,7 +170,14 @@ protected PMML4Result executePMMLLegacy(Context context) {
...
-    protected PMML4Result executePMMLTrusty() {
-        return PMMLCommandExecutorFactory.get().newPMMLCommandExecutor().execute(requestData);
+    protected PMML4Result executePMMLTrusty(Context context) {
+        RegistryContext registryContext = (RegistryContext) context;
+        PMML4Result toReturn = PMMLCommandExecutorFactory.get().newPMMLCommandExecutor().execute(requestData, context);
+        Optional<ExecutionResultImpl> execRes = Optional.ofNullable(registryContext.lookup(ExecutionResultImpl.class));
+        registryContext.register(PMML4Result.class, toReturn);
+        execRes.ifPresent(result -> {
+            result.setResult("results", toReturn);
+        });
```
Comment (danielezonca) [inline drools-core/src/main/java/org/drools/core/command/runtime/pmml/ApplyPmmlModelCommand.java]: Please add a comment like "configure result to be equivalent of old kie-pmml impl"

File: kie-pmml-trusty/kie-pmml-compiler/kie-pmml-compiler-commons/pom.xml [hunk 2]
```diff
@@ -47,11 +56,11 @@
...
-    <dependency>
-      <groupId>org.kie</groupId>
-      <artifactId>kie-memory-compiler</artifactId>
-      <scope>test</scope>
-    </dependency>
+<!--    <dependency>-->
+<!--      <groupId>org.kie</groupId>-->
+<!--      <artifactId>kie-memory-compiler</artifactId>-->
+<!--      <scope>test</scope>-->
+<!--    </dependency>-->
```
Comment (danielezonca) [inline kie-pmml-trusty/kie-pmml-compiler/kie-pmml-compiler-commons/pom.xml]: To be removed

File: kie-pmml-trusty/kie-pmml-evaluator/kie-pmml-evaluator-assembler/src/main/java/org/kie/pmml/evaluator/assembler/command/PMMLCommandExecutorImpl.java [hunk 3]
```diff
@@ -60,7 +67,22 @@ private PMML4Result evaluate(final PMMLRequestData pmmlRequestData, final PMMLRu
...
     }
 
-    private PMMLRuntime getPMMLRuntime(String pmmlFileName) {
-        return PMML_RUNTIME_FACTORY.getPMMLRuntimeFromClasspath(pmmlFileName);
+    private PMMLRuntime getPMMLRuntime(String pmmlFileName, String pmmlModelName, KieBase kieBase) {
+        return PMML_RUNTIME_FACTORY.getPMMLRuntimeFromFileNameModelNameAndKieBase(pmmlFileName, pmmlModelName, kieBase);
+    }
+
+    @SuppressWarnings("rawtype")
+    private PMMLRequestData getCleanedRequestData(PMMLRequestData toClean) {
```
Comment (danielezonca) [inline kie-pmml-trusty/kie-pmml-evaluator/kie-pmml-evaluator-assembler/src/main/java/org/kie/pmml/evaluator/assembler/command/PMMLCommandExecutorImpl.java]: Please add a Javadoc to refer [JSONMarshallerPMMLParamInfo](https://github.com/kiegroup/droolsjbpm-integration/blob/master/kie-server-parent/kie-server-api/src/main/java/org/kie/server/api/marshalling/json/JSONMarshallerPMMLParamInfo.java) behavior

File: kie-memory-compiler/src/main/java/org/kie/memorycompiler/KieMemoryCompiler.java [hunk 4]
```diff
@@ -29,14 +29,39 @@
...
     /**
-     * Compile the given sources and add compiled classes to the given <code>ClassLoader</code>
+     * Compile the given sources <b>without</b> adding compiled classes to the given <code>ClassLoader</code>
      * <b>classNameSourceMap</b>' key must be the <b>FQDN</b> of the class to compile
      *
      * @param classNameSourceMap
      * @param classLoader
      * @return
      */
     public static Map<String, Class<?>> compile(Map<String, String> classNameSourceMap, ClassLoader classLoader) {
```
Comment (danielezonca) [inline kie-memory-compiler/src/main/java/org/kie/memorycompiler/KieMemoryCompiler.java:38]: Please accept this PR to make `compile` method more flexible
https://github.com/gitgabrio/drools/pull/1

File: drools-core/src/main/java/org/drools/core/command/runtime/pmml/ApplyPmmlModelCommand.java (hunk 1 above)
Comment (gitgabrio) [reply drools-core/src/main/java/org/drools/core/command/runtime/pmml/ApplyPmmlModelCommand.java]: @danielezonca 
done

File: kie-pmml-trusty/kie-pmml-evaluator/kie-pmml-evaluator-assembler/src/main/java/org/kie/pmml/evaluator/assembler/command/PMMLCommandExecutorImpl.java (hunk 3 above)
Comment (gitgabrio) [reply kie-pmml-trusty/kie-pmml-evaluator/kie-pmml-evaluator-assembler/src/main/java/org/kie/pmml/evaluator/assembler/command/PMMLCommandExecutorImpl.java]: @danielezonca
done

**Review by danielezonca — APPROVED** [review]

File: drools-core/src/main/java/org/drools/core/command/runtime/pmml/ApplyPmmlModelCommand.java [hunk 5]
```diff
@@ -167,7 +170,14 @@ protected PMML4Result executePMMLLegacy(Context context) {
...
     }
 
-    protected PMML4Result executePMMLTrusty() {
-        return PMMLCommandExecutorFactory.get().newPMMLCommandExecutor().execute(requestData);
+    protected PMML4Result executePMMLTrusty(Context context) {
+        RegistryContext registryContext = (RegistryContext) context;
+        PMML4Result toReturn = PMMLCommandExecutorFactory.get().newPMMLCommandExecutor().execute(requestData, context);
+        // Needed to update the ExecutionResultImpl and the Registry context,
+        // has done inside legacy implementation
+        Optional<ExecutionResultImpl> execRes = Optional.ofNullable(registryContext.lookup(ExecutionResultImpl.class));
```
Comment (mariofusco) [inline drools-core/src/main/java/org/drools/core/command/runtime/pmml/ApplyPmmlModelCommand.java:177]: This use of `Optional` is very questionable. I don't see any valid reason to wrap the result of that lookup into an `Optional` only to test if it is empty or not and take some action accordingly.

Comment (gitgabrio) [reply drools-core/src/main/java/org/drools/core/command/runtime/pmml/ApplyPmmlModelCommand.java:177]: @mariofusco 
This snippet is carbon-copied from [ApplyPmmlModelCommandExecutorImpl](https://github.com/kiegroup/drools/blob/1136d11910cd73aa7cfdf1a9df20aa904c5cbf26/drools-ruleunit/src/main/java/org/drools/ruleunit/command/pmml/ApplyPmmlModelCommandExecutorImpl.java#L199).

Comment (mariofusco) [reply drools-core/src/main/java/org/drools/core/command/runtime/pmml/ApplyPmmlModelCommand.java:177]: I discussed this with @danielezonca and @lucamolteni and they are both happy with this usage of Optional (that continues to seem quite odd to me). Feel free to ignore my former comment.

**Review by mariofusco — APPROVED** [review]

File: kie-pmml-trusty/kie-pmml-compiler/kie-pmml-compiler-commons/pom.xml [hunk 6]
```diff
@@ -20,6 +20,7 @@
       <groupId>org.kie</groupId>
       <artifactId>kie-pmml-compiler-api</artifactId>
     </dependency>
+    <!-- DROOLS -->
```
Comment (jiripetrlik) [inline kie-pmml-trusty/kie-pmml-compiler/kie-pmml-compiler-commons/pom.xml:23]: Is this comment necessary?

File: kie-pmml-trusty/kie-pmml-evaluator/kie-pmml-evaluator-assembler/src/main/java/org/kie/pmml/evaluator/assembler/factories/PMMLRuntimeFactoryInternal.java [hunk 7]
```diff
@@ -47,8 +49,10 @@ private PMMLRuntimeFactoryInternal() {
     }
 
     public static PMMLRuntime getPMMLRuntime(final File pmmlFile) {
-        KnowledgeBuilderImpl kbuilderImpl = (KnowledgeBuilderImpl) KnowledgeBuilderFactory.newKnowledgeBuilder();
-        return getPMMLRuntime(pmmlFile, kbuilderImpl);
+        KieBase kieBase = createKieBase(pmmlFile);
+        return getPMMLRuntime(kieBase);
+//        KnowledgeBuilderImpl kbuilderImpl = (KnowledgeBuilderImpl) KnowledgeBuilderFactory.newKnowledgeBuilder();
```
Comment (jiripetrlik) [inline kie-pmml-trusty/kie-pmml-evaluator/kie-pmml-evaluator-assembler/src/main/java/org/kie/pmml/evaluator/assembler/factories/PMMLRuntimeFactoryInternal.java]: Please remove commented code.

File: kie-pmml-trusty/kie-pmml-evaluator/kie-pmml-evaluator-assembler/src/test/java/org/kie/pmml/evaluator/assembler/implementations/HasKnowledgeBuilderImplTest.java [hunk 8]
```diff
@@ -0,0 +1,70 @@
...
+        Class<?> retrieved = hasKnowledgeBuilder.compileAndLoadClass(sourcesMap, CLASS_SOURCE_NAME);
+        assertNotNull(retrieved);
+        assertEquals(CLASS_SOURCE_NAME, retrieved.getName());
+    }
+
+    @Test
+    public void compileAndLoadClassMultipleTimes() {
+        Map<String, String> sourcesMap = new HashMap<>();
+        sourcesMap.put(CLASS_SOURCE_NAME, CLASS_SOURCE);
+        IntStream.range(0,3).forEach(value -> hasKnowledgeBuilder.compileAndLoadClass(sourcesMap, CLASS_SOURCE_NAME));
```
Comment (jiripetrlik) [inline kie-pmml-trusty/kie-pmml-evaluator/kie-pmml-evaluator-assembler/src/test/java/org/kie/pmml/evaluator/assembler/implementations/HasKnowledgeBuilderImplTest.java]: It seems a little bit suspicious that there is no assert in this test.

**Review by jiripetrlik — CHANGES_REQUESTED** [review]
I have few comments. It would be also good to improve test coverage. It seems that "PMMLRuntimeFactoryImpl" and "PMMLRuntimeFactoryInternal" does not have a good test coverage.

**Comment by Ginxo** [comment]
Jenkins retest this

**Comment by danielezonca** [comment]
jenkins run cdb

**Review by jiripetrlik — APPROVED** [review]
@gitgabrio
Thank you for the changes and improved test coverage. It looks good to me now.

**Comment by gitgabrio** [comment]
@danielezonca @jiripetrlik 
I'm adding further tests to better cover the PMMLRuntime factories

**Comment by danielezonca** [comment]
> @danielezonca @jiripetrlik
> I'm adding further tests to better cover the PMMLRuntime factories

Merged because of 7.49 cutoff, please create a new PR to improve coverage

**Comment by gitgabrio** [comment]
:+1:

**Pull request merged**